lib/Refactoring/AddExplicitCodableImplementation.cpp

```diff
@@ -12,183 +12,230 @@
-    Indent = (CurrentIndent + ExtraIndent).str();
-  } else {
-    Indent = CurrentIndent.str();
+public:
+  static AddCodableContext getFromCursorInfo(ResolvedCursorInfoPtr Info);
+
+  bool isApplicable() const {
+    if (!IDC)
+      return false;
+
+    // Check if 'IDC' conforms to 'Encodable' and/or 'Decodable' and any of the
+    // requirements are synthesized.
+    for (auto *conformance : IDC->getLocalConformances()) {
+      auto protocol = conformance->getProtocol();
+      auto kind = protocol->getKnownProtocolKind();
+      if (kind == KnownProtocolKind::Encodable ||
+          kind == KnownProtocolKind::Decodable) {
+        // Check if any of the protocol requirements are synthesized.
+        for (auto requirement : protocol->getProtocolRequirements()) {
+          auto witness = conformance->getWitnessDecl(requirement);
+          if (!witness || witness->isSynthesized())
+            return true;
+        }
+      }
+    }
+    return false;
   }
 
-  ExtraIndentStreamPrinter Printer(OS, Indent);
-  Printer.printNewline();
-  SynthesizedCodablePrinter Walker(Printer);
-  DC->getAsDecl()->walk(Walker);
-}
+  void getInsertion(SourceLoc &insertLoc, std::string &insertText) const {
+    insertLoc = getInsertStartLoc();
+    llvm::raw_string_ostream OS(insertText);
+    printInsertText(OS);
+  }
+};
 
 AddCodableContext
-AddCodableContext::getDeclarationContextFromInfo(ResolvedCursorInfoPtr Info) {
+AddCodableContext::getFromCursorInfo(ResolvedCursorInfoPtr Info) {
   auto ValueRefInfo = dyn_cast<ResolvedValueRefCursorInfo>(Info);
   if (!ValueRefInfo) {
-    return AddCodableContext();
+    return nullptr;
   }
+
+  if (auto *ext = ValueRefInfo->getExtTyRef()) {
+    // For 'extension Outer.Inner: Codable {}', only 'Inner' part is valid.
+    if (ext->getExtendedNominal() == ValueRefInfo->getValueD()) {
+      return AddCodableContext(ext);
+    } else {
+      return nullptr;
+    }
+  }
+
   if (!ValueRefInfo->isRef()) {
-    if (auto *NomDecl = dyn_cast<NominalTypeDecl>(ValueRefInfo->getValueD())) {
-      return AddCodableContext(NomDecl);
+    if (auto *nominal = dyn_cast<NominalTypeDecl>(ValueRefInfo->getValueD())) {
+      return AddCodableContext(nominal);
     }
   }
-  // TODO: support extensions
-  // (would need to get synthesized nodes from the main decl,
-  // and only if it's in the same file?)
-  return AddCodableContext();
+
+  return nullptr;
 }
 } // namespace
 
 bool RefactoringActionAddExplicitCodableImplementation::isApplicable(
     ResolvedCursorInfoPtr Tok, DiagnosticEngine &Diag) {
-  return AddCodableContext::getDeclarationContextFromInfo(Tok).isValid();
+  return AddCodableContext::getFromCursorInfo(Tok).isApplicable();
 }
 
 bool RefactoringActionAddExplicitCodableImplementation::performChange() {
-  auto Context = AddCodableContext::getDeclarationContextFromInfo(CursorInfo);
+  auto Context = AddCodableContext::getFromCursorInfo(CursorInfo);
+  assert(Context.isApplicable());
```

Comment:
`assert(Context.isApplicable() && "Should not run performChange when refactoring is not applicable")` would make this super clear :D.